File: app/lib/sketch.dart
/// Screen-space geometry for the drawn sketch.
///
/// Kept free of Flutter widgets and of the map so it can be unit tested without
/// a device. Everything here works in screen pixels; conversion to latitude and
/// longitude happens once, at the end of a stroke.
library;

import 'dart:math' as math;
import 'dart:ui' show Offset;

/// Thin a freehand stroke with the Douglas-Peucker algorithm.
///
/// A finger emits hundreds of points for a shape that needs tens. Thinning
/// matters here for a specific reason: every surviving point costs one platform
/// channel round trip when screen coordinates are converted to geographic ones,
/// so an unthinned stroke turns a cheap conversion into a visible stall.
List<Offset> simplify(List<Offset> points, double tolerance) {
  if (points.length < 3 || tolerance <= 0) return List.of(points);
  final out = <Offset>[];
  _douglasPeucker(points, 0, points.length - 1, tolerance, out);
  out.add(points.last);
  return out;
}

void _douglasPeucker(
  List<Offset> pts,
  int first,
  int last,
  double tolerance,
  List<Offset> out,
) {
  var worst = 0.0;
  var worstIndex = 0;

  for (var i = first + 1; i < last; i++) {
    final d = _distanceToSegment(pts[i], pts[first], pts[last]);
    if (d > worst) {
      worst = d;
      worstIndex = i;
    }
  }

  if (worst > tolerance) {
    _douglasPeucker(pts, first, worstIndex, tolerance, out);
    _douglasPeucker(pts, worstIndex, last, tolerance, out);
  } else {
    out.add(pts[first]);
  }
}

double _distanceToSegment(Offset p, Offset a, Offset b) {
  final dx = b.dx - a.dx;
  final dy = b.dy - a.dy;
  final lengthSquared = dx * dx + dy * dy;

  if (lengthSquared <= 1e-12) return (p - a).distance;

  final t = (((p.dx - a.dx) * dx + (p.dy - a.dy) * dy) / lengthSquared)
      .clamp(0.0, 1.0);
  return (p - Offset(a.dx + dx * t, a.dy + dy * t)).distance;
}

/// Total length of a stroke in pixels.
double strokeLength(List<Offset> points) {
  var total = 0.0;
  for (var i = 1; i < points.length; i++) {
    total += (points[i] - points[i - 1]).distance;
  }
  return total;
}

/// Whether a stroke is long enough to be a deliberate line rather than a tap.
///
/// Without this, every stray tap on the map would clear the drawn route.
bool isDeliberateStroke(List<Offset> points) =>
    points.length >= 3 && strokeLength(points) >= 24.0;

/// Smallest rectangle containing the stroke, or null when there is nothing.
math.Rectangle<double>? strokeBounds(List<Offset> points) {
  if (points.isEmpty) return null;
  var minX = points.first.dx, maxX = points.first.dx;
  var minY = points.first.dy, maxY = points.first.dy;
  for (final p in points) {
    minX = math.min(minX, p.dx);
    maxX = math.max(maxX, p.dx);
    minY = math.min(minY, p.dy);
    maxY = math.max(maxY, p.dy);
  }
  return math.Rectangle<double>(minX, minY, maxX - minX, maxY - minY);
}

File: web/src/sketch.ts
/**
 * Screen-space geometry for the drawn sketch.
 *
 * Kept free of the DOM and of the map so it can be unit tested without a
 * browser. Everything here works in CSS pixels; conversion to latitude and
 * longitude happens once, at the end of a stroke.
 *
 * This is a port of the Flutter client's `sketch.dart`, and the tests came with
 * it. The thinning here is a cheap first pass in pixel space: the thinning that
 * decides route quality is done in metres, inside the Rust engine, where a
 * tolerance has a meaning that survives a change of zoom level.
 */

export interface Point {
  readonly x: number;
  readonly y: number;
}

export interface Bounds {
  readonly left: number;
  readonly top: number;
  readonly right: number;
  readonly bottom: number;
}

/**
 * Thin a freehand stroke with the Douglas-Peucker algorithm.
 *
 * A finger emits hundreds of points for a shape that needs tens. Thinning here
 * keeps the live path element small enough to re-render at sixty frames per
 * second while the finger is still moving.
 */
export function simplify(points: readonly Point[], tolerance: number): Point[] {
  if (points.length < 3 || tolerance <= 0) return [...points];
  const out: Point[] = [];
  douglasPeucker(points, 0, points.length - 1, tolerance, out);
  out.push(points[points.length - 1]);
  return out;
}

function douglasPeucker(
  pts: readonly Point[],
  first: number,
  last: number,
  tolerance: number,
  out: Point[],
): void {
  let worst = 0;
  let worstIndex = 0;

  for (let i = first + 1; i < last; i++) {
    const d = distanceToSegment(pts[i], pts[first], pts[last]);
    if (d > worst) {
      worst = d;
      worstIndex = i;
    }
  }

  if (worst > tolerance) {
    douglasPeucker(pts, first, worstIndex, tolerance, out);
    douglasPeucker(pts, worstIndex, last, tolerance, out);
  } else {
    out.push(pts[first]);
  }
}

function distanceToSegment(p: Point, a: Point, b: Point): number {
  const dx = b.x - a.x;
  const dy = b.y - a.y;
  const lengthSquared = dx * dx + dy * dy;

  if (lengthSquared <= 1e-12) return Math.hypot(p.x - a.x, p.y - a.y);

  const t = Math.min(
    1,
    Math.max(0, ((p.x - a.x) * dx + (p.y - a.y) * dy) / lengthSquared),
  );
  return Math.hypot(p.x - (a.x + dx * t), p.y - (a.y + dy * t));
}

/** Total length of a stroke in pixels. */
export function strokeLength(points: readonly Point[]): number {
  let total = 0;
  for (let i = 1; i < points.length; i++) {
    total += Math.hypot(points[i].x - points[i - 1].x, points[i].y - points[i - 1].y);
  }
  return total;
}

/**
 * Whether a stroke is long enough to be a deliberate line rather than a tap.
 *
 * Without this, every stray touch on the map would wipe a drawn route.
 */
export function isDeliberateStroke(points: readonly Point[]): boolean {
  return points.length >= 3 && strokeLength(points) >= 24;
}

/** Smallest rectangle containing the stroke, or null when there is nothing. */
export function strokeBounds(points: readonly Point[]): Bounds | null {
  if (points.length === 0) return null;
  let left = points[0].x;
  let right = points[0].x;
  let top = points[0].y;
  let bottom = points[0].y;
  for (const p of points) {
    left = Math.min(left, p.x);
    right = Math.max(right, p.x);
    top = Math.min(top, p.y);
    bottom = Math.max(bottom, p.y);
  }
  return { left, top, right, bottom };
}

/** An SVG path `d` attribute for a stroke, or empty when there is nothing to draw. */
export function toPathData(points: readonly Point[]): string {
  if (points.length < 2) return '';
  let d = `M ${points[0].x.toFixed(1)} ${points[0].y.toFixed(1)}`;
  for (let i = 1; i < points.length; i++) {
    d += ` L ${points[i].x.toFixed(1)} ${points[i].y.toFixed(1)}`;
  }
  return d;
}
